### DISCUSSION TITLE
Not able to read and decrypt private keys

### DISCUSSION THREAD

**Discussion by Tabish13** [openpgpjs/openpgpjs#1431] [Q&A]
- OpenPGP.js version: 5.0.0
- Affected platform (Node.js version): v14.17.6

I am having already generated pgp private key which is a valid key trying to encrypt a message using following code
```js
openpgp.config.prefer_hash_algorithm = openpgp.enums.hash.sha256;
openpgp.config.encryption_cipher = openpgp.enums.symmetric.aes256;
openpgp.config.compression = openpgp.enums.compression.zip;

var payloadBody=JSON.stringify({"header": {
    "msgId": "2021101101",
   
           "orgId": "data",
   
           "timestamp": "2021-10-20T10:46:30.552+05:30"
   
       },
       "data": {
           "phoneNumber": "11111"
       }
   });
console.log("");
console.log(payloadBody);
(async ()=>{

    const message = await openpgp.createMessage({ text: payloadBody });
    const config = {};
    const privateKey = await openpgp.readPrivateKey({ armoredKey: privateKey});

    console.log(privateKey);
    const privateKeyPass = await openpgp.decryptKey({
        privateKey: privateKey,
        passphrase
    });
    
    const encrypted = await openpgp.encrypt({
        message, // input as Message object
        encryptionKeys: await openpgp.readKey({ armoredKey: publicKey }),
        signingKeys: privateKeyPass,
        format: 'armored',
    
       
    });
    console.log(encrypted)

})();
```

getting error in decrypting the key i.e. Error: Error decrypting private key: Key packet is already decrypted.
And if I dont decrypt and encrypt the payload I am getting Error encrypting message: Signing key is not decrypted.

**Comment by larabr** [comment]
It looks like your key is only partially encrypted. 
You can do the following to encrypt all the packets:
```js
await Promise.all(
  privateKey.getKeys().map(async ({keyPacket}) => { 
    if (keyPacket.isDecrypted()) {
      await keyPacket.encrypt(passphrase);
      keyPacket.clearPrivateParams();
    }
  });
)
```
then try again to decrypt `privateKey` using `decryptKey`.

PS: when setting `openpgp.config`, you are using the v4 configuration options instead of the v5 ones, so your settings do not apply. The valid v5 config options are here: https://github.com/openpgpjs/openpgpjs/blob/master/src/config/config.js .

**Reply by Tabish13** [reply]
thanks for the config have changed it as per the v5.
I have tried the above code as I am using only readPrivateKey and not keys getting removed the map getting keyPacket directly from the readPrivateKey and on trying  keyPacket.isDecrypted() getting the value as false.
any other possible reason why the private key decryption is not possible same pgp key works with java code for encryptions.

**Reply by Tabish13** [reply]
Thanks for the guide I have solved this by doing below  
`privateKey.keyPacket.decrypt(passphrase);` instead of passing the readKey to decryptkey function.

**Reply by larabr** [reply]
The code I posted before is supposed to be run after `const privateKey = await openpgp.readPrivateKey({ armoredKey: 
 armoredPrivateKey})`, since `privateKey` can still contain subkeys (`readPrivateKeys` has a different use case).

Calling `keyPacket.decrypt()` is not recommended because -- aside from potentially leaving the key in half-decrypted state -- it does not properly check the integrity of the key, so you should at least call `await keyPacket.validate()` afterwards.
But it's really best you decrypt your key using `decryptKey`.

**Reply by Tabish13** [reply]
The above approach allows me to decrypt the key and also able to encrypt the payload as expected
If I am using the previous code `if (keyPacket.isDecrypted()) {
      await keyPacket.encrypt(passphrase);
      keyPacket.clearPrivateParams();
    }`
    
The if conditon never get executed as it return false but I get error  Error: Key is not decrypted on using 
`await keyPacket.validate()` 
I get the same error again.
   let me know if there is a cleaner approach to decrypt it using decryptKey resolving the error of key is not decrypted 
   Below is the keyPacket received on reading the private key
 
  ```
{
  keyPacket: xo {
    version: 4,
    created: 2021-09-13T05:04:06.000Z,
    algorithm: 'rsaEncryptSign',
    publicParams: { n: [Uint8Array], e: [Uint8Array] },
    expirationTimeV3: 0,
    fingerprint: Uint8Array(20) [
      ..
    ],
    keyID: xe { bytes: '..' },
    keyMaterial: Uint8Array(988) [
       ..
    ],
    isEncrypted: true,
    s2kUsage: 254,
    s2k: vo {
      algorithm: 'sha1',
      type: 'iterated',
      c: 206,
      salt: [Uint8Array]
    },
    symmetric: 'aes128',
    aead: null,
    privateParams: null,
    packets: no(0) [],
    fromStream: false,
    iv: Uint8Array(16) [
    ...
    ]
  }


```

**Reply by larabr** [reply]
I don't understand if you've actually tried running my code or not, because it works for me -- if it doesn't work for you, it means the key has a different issue.

The key packet you have shown is encrypted, so we don't want to call `encrypt()` on it on purpose. But there are other key packets in that key. Can you check `privateKey.subkeys.map(subkey => subkey.keyPacket)` ?

**Reply by Tabish13** [reply]
Had added few lines previously which could have caused the issue
Thank you @larabr your code worked
```(async ()=>{
    const message = await openpgp.createMessage({ text: payloadBody });
    const privateKey = await openpgp.readPrivateKey({ armoredKey: keys.privateKey}); 
    console.log(privateKey);
    await Promise.all(
        privateKey.getKeys().map(async ({keyPacket}) => { 
          if (keyPacket.isDecrypted()) {
            await keyPacket.encrypt(passphrase);
            keyPacket.clearPrivateParams();
          }
        })
      );
    const privateKeyPass = await openpgp.decryptKey({
        privateKey: privateKey,
        passphrase
    });
    const encrypted = await openpgp.encrypt({
        message, // input as Message object
        encryptionKeys: await openpgp.readKey({ armoredKey: keys.publicKey }),
        signingKeys: privateKeyPass,
        format: 'armored',
    });
    console.log(encrypted);
    
})();```
